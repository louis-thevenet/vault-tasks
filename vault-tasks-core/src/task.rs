use chrono::{NaiveDate, NaiveDateTime};
use color_eyre::{eyre::bail, Result};
use core::fmt;
use std::{
    cmp::Ordering,
    fmt::Display,
    fs::{read_to_string, File},
    io::Write,
    path::PathBuf,
};
use tracing::{debug, info};

use crate::TasksConfig;

const STATE_TO_DO_EMOJI: &str = "‚ùå";
const STATE_DONE_EMOJI: &str = "‚úÖ";
const DUE_DATE_EMOJI: &str = "üìÖ";
pub const PRIORITY_EMOJI: &str = "‚ùó";
pub const TODAY_FLAG_EMOJI: &str = "‚òÄÔ∏è";

/// A task's state
/// Ordering is `Todo < Done`
#[derive(Debug, Hash, Eq, PartialEq, Clone, PartialOrd, Ord)]
pub enum State {
    ToDo,
    Done,
}
impl Display for State {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Done => write!(f, "{STATE_DONE_EMOJI}")?,
            Self::ToDo => write!(f, "{STATE_TO_DO_EMOJI}")?,
        }
        Ok(())
    }
}
#[derive(Debug, Hash, Eq, PartialEq, Clone)]
/// This type accounts for the case where the task has a due date but no exact due time
pub enum DueDate {
    NoDate,
    Day(NaiveDate),
    DayTime(NaiveDateTime),
}
impl Display for DueDate {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Day(date) => write!(f, "{date}"),
            Self::DayTime(date) => write!(f, "{date}"),
            Self::NoDate => Ok(()),
        }
    }
}

impl DueDate {
    pub fn to_display_format(&self, not_american_format: bool) -> String {
        if matches!(self, Self::NoDate) {
            String::new()
        } else {
            format!(
                "{DUE_DATE_EMOJI} {}",
                self.to_string_format(not_american_format)
            )
        }
    }
    pub fn to_string_format(&self, not_american_format: bool) -> String {
        let format_date = if not_american_format {
            "%d/%m/%Y"
        } else {
            "%Y/%m/%d"
        };
        let format_datetime = if not_american_format {
            "%d/%m/%Y %T"
        } else {
            "%Y/%m/%d %T"
        };

        match self {
            Self::Day(date) => date.format(format_date).to_string(),
            Self::DayTime(date) => date.format(format_datetime).to_string(),
            Self::NoDate => String::new(),
        }
    }

    pub fn get_relative_str(&self) -> Option<String> {
        let now = chrono::Local::now();
        let time_delta = match self {
            Self::NoDate => return None,
            Self::Day(date) => now.date_naive().signed_duration_since(*date),
            Self::DayTime(date_time) => {
                now.date_naive().signed_duration_since(date_time.date())
                    + now.time().signed_duration_since(date_time.time())
            }
        };

        let (prefix, suffix) = match time_delta.num_seconds().cmp(&0) {
            Ordering::Less => (String::from("in "), String::new()),
            Ordering::Equal => (String::new(), String::new()),
            Ordering::Greater => (String::new(), String::from(" ago")),
        };

        let time_delta_abs = time_delta.abs();

        if time_delta_abs.is_zero() {
            return Some(String::from("today"));
        }
        if time_delta.num_seconds() < 0 && time_delta_abs.num_days() == 1 {
            return Some(String::from("tomorrow"));
        }
        if time_delta.num_seconds() > 0 && time_delta_abs.num_days() == 1 {
            return Some(String::from("yesterday"));
        }

        let res = if 4 * 12 * 2 <= time_delta_abs.num_weeks() {
            format!("{} years", time_delta_abs.num_weeks() / (12 * 4))
        } else if 5 <= time_delta_abs.num_weeks() {
            format!("{} months", time_delta_abs.num_weeks() / 4)
        } else if 2 <= time_delta_abs.num_weeks() {
            format!("{} weeks", time_delta_abs.num_weeks())
        } else if 2 <= time_delta_abs.num_days() {
            format!("{} days", time_delta_abs.num_days())
        } else {
            format!("{} hours", time_delta_abs.num_hours())
        };
        Some(format!("{prefix}{res}{suffix}"))
    }
}

#[derive(Debug, Hash, Eq, PartialEq, Clone)]
pub struct Task {
    pub subtasks: Vec<Task>,
    pub description: Option<String>,
    pub due_date: DueDate,
    pub filename: String,
    pub line_number: usize,
    pub name: String,
    pub priority: usize,
    pub state: State,
    pub tags: Option<Vec<String>>,
    pub is_today: bool,
}

impl Default for Task {
    fn default() -> Self {
        Self {
            due_date: DueDate::NoDate,
            name: String::new(),
            priority: 0,
            state: State::ToDo,
            tags: None,
            description: None,
            line_number: 1,
            subtasks: vec![],
            filename: String::new(),
            is_today: false,
        }
    }
}

impl fmt::Display for Task {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let state = self.state.to_string();
        let title = format!("{state} {}", self.name);
        writeln!(f, "{title}")?;

        let mut data_line = String::new();
        let is_today = if self.is_today {
            format!("{TODAY_FLAG_EMOJI} ")
        } else {
            String::new()
        };
        data_line.push_str(&is_today);
        let due_date_str = self.due_date.to_string();

        if !due_date_str.is_empty() {
            data_line.push_str(&format!(
                "{DUE_DATE_EMOJI} {due_date_str} ({})",
                self.due_date.get_relative_str().unwrap_or_default()
            ));
        }
        if self.priority > 0 {
            data_line.push_str(&format!("{}{} ", PRIORITY_EMOJI, self.priority));
        }
        if !data_line.is_empty() {
            writeln!(f, "{data_line}")?;
        }
        let mut tag_line = String::new();
        if self.tags.is_some() {
            tag_line.push_str(
                &self
                    .tags
                    .clone()
                    .unwrap()
                    .iter()
                    .map(|t| format!("#{t}"))
                    .collect::<Vec<String>>()
                    .join(" "),
            );
        }
        if !tag_line.is_empty() {
            writeln!(f, "{tag_line}")?;
        }
        if let Some(description) = self.description.clone() {
            for l in description.lines() {
                writeln!(f, "{l}")?;
            }
        }
        Ok(())
    }
}
impl Task {
    pub fn get_fixed_attributes(&self, config: &TasksConfig, indent_length: usize) -> String {
        let indent = " ".repeat(indent_length);

        let state_str = match self.state {
            State::Done => config.task_state_marker,
            State::ToDo => ' ',
        };

        let priority = if self.priority > 0 {
            format!("p{} ", self.priority)
        } else {
            String::new()
        };

        let mut due_date = self.due_date.to_string_format(!config.use_american_format);
        if !due_date.is_empty() {
            due_date.push(' ');
        }

        let tags_str = self.tags.as_ref().map_or_else(String::new, |tags| {
            tags.clone()
                .iter()
                .map(|t| format!("#{t}"))
                .collect::<Vec<String>>()
                .join(" ")
        });

        let today_tag = if self.is_today {
            String::from(" @today")
        } else {
            String::new()
        };

        let res = format!(
            "{}- [{}] {} {}{}{}{}",
            indent, state_str, self.name, due_date, priority, tags_str, today_tag
        );
        res.trim_end().to_string()
    }

    pub fn fix_task_attributes(&self, config: &TasksConfig, path: &PathBuf) -> Result<()> {
        let content = read_to_string(path.clone())?;
        let mut lines = content.split('\n').collect::<Vec<&str>>();

        if lines.len() < self.line_number - 1 {
            bail!(
                "Task's line number {} was greater than length of file {:?}",
                self.line_number,
                path
            );
        }

        let indent_length = lines[self.line_number - 1]
            .chars()
            .take_while(|c| c.is_whitespace())
            .count();

        let fixed_line = self.get_fixed_attributes(config, indent_length);

        if lines[self.line_number - 1] != fixed_line {
            debug!(
                "\nReplacing\n{}\nWith\n{}\n",
                lines[self.line_number - 1],
                self.get_fixed_attributes(config, indent_length,)
            );
            lines[self.line_number - 1] = &fixed_line;

            let mut file = File::create(path)?;
            file.write_all(lines.join("\n").as_bytes())?;

            info!("Wrote to {path:?} at line {}", self.line_number);
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests_tasks {
    use chrono::NaiveDate;
    use pretty_assertions::assert_eq;

    use crate::{
        task::{DueDate, State, Task},
        TasksConfig,
    };

    #[test]
    fn test_fix_attributes() {
        let config = TasksConfig {
            use_american_format: true,
            ..Default::default()
        };
        let task = Task {
            due_date: DueDate::Day(NaiveDate::from_ymd_opt(2021, 12, 3).unwrap()),
            name: String::from("Test Task"),
            priority: 1,
            state: State::ToDo,
            tags: Some(vec![String::from("tag1"), String::from("tag2")]),
            description: Some(String::from("This is a test task.")),
            line_number: 2,
            ..Default::default()
        };
        let res = task.get_fixed_attributes(&config, 0);
        assert_eq!(res, "- [ ] Test Task 2021/12/03 p1 #tag1 #tag2");
    }

    #[test]
    fn test_fix_attributes_with_no_date() {
        let config = TasksConfig {
            task_state_marker: 'x',
            ..Default::default()
        };
        let task = Task {
            due_date: DueDate::NoDate,
            name: String::from("Test Task with No Date"),
            priority: 2,
            state: State::Done,
            tags: Some(vec![String::from("tag3")]),
            description: None,
            line_number: 3,
            ..Default::default()
        };

        let res = task.get_fixed_attributes(&config, 0);
        assert_eq!(res, "- [x] Test Task with No Date p2 #tag3");
    }
    #[test]
    fn test_fix_attributes_with_today_tag() {
        let config = TasksConfig {
            task_state_marker: 'x',
            ..Default::default()
        };
        let task = Task {
            due_date: DueDate::NoDate,
            name: String::from("Test Task with Today tag"),
            priority: 2,
            state: State::Done,
            tags: Some(vec![String::from("tag3")]),
            description: None,
            line_number: 3,
            is_today: true,
            ..Default::default()
        };

        let res = task.get_fixed_attributes(&config, 0);
        assert_eq!(res, "- [x] Test Task with Today tag p2 #tag3 @today");
    }
}
#[cfg(test)]
mod tests_due_date {
    use chrono::TimeDelta;

    use crate::task::DueDate;

    #[test]
    fn test_relative_date() {
        let now = chrono::Local::now();

        let tests = vec![
            (-1, "yesterday"),
            (0, "today"),
            (1, "tomorrow"),
            (7, "in 7 days"),
            (17, "in 2 weeks"),
            (65, "in 2 months"),
            (800, "in 2 years"),
        ];
        for (days, res) in tests {
            let due_date = DueDate::Day(
                now.checked_add_signed(TimeDelta::days(days))
                    .unwrap()
                    .date_naive(),
            );
            assert_eq!(due_date.get_relative_str(), Some(String::from(res)));
        }
    }
}
