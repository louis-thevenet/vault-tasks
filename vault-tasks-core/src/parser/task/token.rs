use chrono::{NaiveDate, NaiveTime};

use crate::task::State;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Token {
    DueDate(NaiveDate),
    DueTime(NaiveTime),
    Name(String),
    Priority(usize),
    Tag(String),
    State(State),
    TodayFlag,
}
