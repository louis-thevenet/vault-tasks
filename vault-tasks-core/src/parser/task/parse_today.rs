use winnow::{
    combinator::{alt, preceded},
    PResult, Parser,
};

use super::token::Token;

/// Parses a `Token::TodayFlag` of the form of the form "@t", @tdy", "@tod" or "@today".
pub fn parse_today(input: &mut &str) -> PResult<Token> {
    preceded('@', alt(("today", "tod", "tdy", "t"))).parse_next(input)?;
    Ok(Token::TodayFlag)
}

#[cfg(test)]
mod tests {
    use crate::parser::task::{parse_today::parse_today, token::Token};

    #[test]
    fn test_parse_today_tag() {
        let mut with_today = "@t";
        assert_eq!(parse_today(&mut with_today), Ok(Token::TodayFlag));
        let mut with_today = "@today";
        assert_eq!(parse_today(&mut with_today), Ok(Token::TodayFlag));
        let mut with_today = "@tdy";
        assert_eq!(parse_today(&mut with_today), Ok(Token::TodayFlag));
    }
    #[test]
    fn test_parse_today_tag_fail() {
        let mut should_fail = "today";
        assert!(parse_today(&mut should_fail).is_err());
    }
}
