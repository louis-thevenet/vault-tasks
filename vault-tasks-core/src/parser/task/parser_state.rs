use winnow::{
    Parser, Result,
    combinator::{delimited, preceded},
    token::any,
};

use crate::config::TaskMarkerConfig;
use crate::task::State;

use super::token::Token;

/// Parses a `TaskState` from an input string.
pub fn parse_task_state(input: &mut &str, task_marker_config: &TaskMarkerConfig) -> Result<Token> {
    match preceded("- ", delimited("[", any, "]")).parse_next(input) {
        Ok(c) => {
            if c == task_marker_config.todo {
                Ok(Token::State(State::ToDo))
            } else if c == task_marker_config.incomplete {
                Ok(Token::State(State::Incomplete))
            } else if c == task_marker_config.canceled {
                Ok(Token::State(State::Canceled))
            } else {
                Ok(Token::State(State::Done))
            }
        }

        Err(error) => Err(error),
    }

    // // This version only supports X to mark tasks done
    // match alt(("- [ ]", "- [X]")).parse_next(input) {
    //     Err(error) => Err(error),
    //     Ok("- [ ]") => Ok(Token::State(TaskState::ToDo)),
    //     _ => Ok(Token::State(TaskState::Done)),
    // }
}
#[cfg(test)]
mod test {
    use crate::{
        config::TaskMarkerConfig,
        parser::task::{parser_state::parse_task_state, token::Token},
        task::State,
    };
    fn config() -> TaskMarkerConfig {
        TaskMarkerConfig {
            done: 'x',
            todo: ' ',
            incomplete: '/',
            canceled: '-',
        }
    }

    #[test]
    fn test_parse_task_state_todo() {
        let mut input = "- [ ]";
        let expected = Ok(Token::State(State::ToDo));
        let config = &config();
        assert_eq!(parse_task_state(&mut input, config), expected);
    }
    #[test]
    fn test_parse_task_state_done() {
        let mut input = "- [X]";
        let expected = Ok(Token::State(State::Done));
        let config = &config();
        assert_eq!(parse_task_state(&mut input, config), expected);
    }
    #[test]
    fn test_parse_task_state_done_alt() {
        let mut input = "- [o]";
        let expected = Ok(Token::State(State::Done));
        let config = &config();
        assert_eq!(parse_task_state(&mut input, config), expected);
    }
    #[test]
    fn test_parse_task_state_canceled() {
        let mut input = "- [-]";
        let expected = Ok(Token::State(State::Canceled));
        let config = &config();
        assert_eq!(parse_task_state(&mut input, config), expected);
    }
    #[test]
    fn test_parse_task_state_fail() {
        let mut input = "- o]";
        let config = &config();
        assert!(parse_task_state(&mut input, config).is_err());
    }
}
