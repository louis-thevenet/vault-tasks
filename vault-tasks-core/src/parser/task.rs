mod parse_today;
mod parser_due_date;
mod parser_priorities;
mod parser_state;
mod parser_tags;
mod parser_time;
mod token;

use chrono::NaiveDateTime;
use parse_today::parse_today;
use parser_due_date::parse_naive_date;
use parser_priorities::parse_priority;
use parser_state::parse_task_state;
use parser_tags::parse_tag;
use parser_time::parse_naive_time;
use token::Token;
use tracing::error;
use winnow::{
    combinator::{alt, fail, repeat},
    token::any,
    PResult, Parser,
};

use crate::{
    task::{DueDate, Task},
    TasksConfig,
};

/// Parses a `Token` from an input string.FileEntry
fn parse_token(input: &mut &str, config: &TasksConfig) -> PResult<Token> {
    alt((
        |input: &mut &str| parse_naive_date(input, config.use_american_format),
        parse_naive_time,
        parse_tag,
        parse_task_state,
        parse_priority,
        parse_today,
        |input: &mut &str| {
            let res = repeat(0.., any)
                .fold(String::new, |mut string, c| {
                    string.push(c);
                    string
                })
                .parse_next(input)?;
            Ok(Token::Name(res))
        },
    ))
    .parse_next(input)
}

/// Parses a `Task` from an input string. Filename must be specified to be added to the task.
///
/// # Errors
///
/// Will return an error if the task can't be parsed.
#[allow(clippy::module_name_repetitions)]
pub fn parse_task(input: &mut &str, filename: String, config: &TasksConfig) -> PResult<Task> {
    let task_state = match parse_task_state(input)? {
        Token::State(state) => Ok(state),
        _ => fail(input),
    }?;

    let mut token_parser = |input: &mut &str| parse_token(input, config);

    let tokens = input
        .split_ascii_whitespace()
        .map(|token| token_parser.parse(token));

    let mut task = Task {
        state: task_state,
        filename,
        ..Default::default()
    };

    // Placeholders for a date and a time
    let mut due_date_opt = None;
    let mut due_time_opt = None;
    let mut name_vec = vec![]; // collects words that aren't tokens from the input string

    for token_res in tokens {
        match token_res {
            Ok(Token::DueDate(date)) => due_date_opt = Some(date),
            Ok(Token::DueTime(time)) => due_time_opt = Some(time),
            Ok(Token::Name(name)) => name_vec.push(name),
            Ok(Token::Priority(p)) => task.priority = p,
            Ok(Token::State(state)) => task.state = state,
            Ok(Token::Tag(tag)) => {
                if let Some(ref mut tags) = task.tags {
                    tags.push(tag);
                } else {
                    task.tags = Some(vec![tag]);
                }
            }
            Ok(Token::TodayFlag) => task.is_today = true,
            Err(error) => error!("Error: {error:?}"),
        }
    }

    if !name_vec.is_empty() {
        task.name = name_vec.join(" ");
    }

    let now = chrono::Local::now();
    let (due_date, has_date) = (
        due_date_opt.unwrap_or_else(|| now.date_naive()),
        due_date_opt.is_some(),
    );
    let (due_time, has_time) = (
        due_time_opt.unwrap_or_else(|| now.time()),
        due_time_opt.is_some(),
    );
    let due_date_time = if has_date {
        if has_time {
            DueDate::DayTime(NaiveDateTime::new(due_date, due_time))
        } else {
            DueDate::Day(due_date)
        }
    } else if has_time {
        DueDate::DayTime(NaiveDateTime::new(now.date_naive(), due_time))
    } else {
        DueDate::NoDate
    };
    task.due_date = due_date_time;
    Ok(task)
}
#[cfg(test)]
mod test {

    use chrono::{Datelike, Days, NaiveDate, NaiveDateTime, NaiveTime};

    use crate::{
        parser::task::parse_task,
        task::{DueDate, State, Task},
        TasksConfig,
    };
    #[test]
    fn test_parse_task_no_description() {
        let mut input = "- [x] 10/15 task_name #done";
        let config = TasksConfig {
            use_american_format: true,
            ..Default::default()
        };
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let year = chrono::Local::now().year();
        let expected = Task {
            name: "task_name".to_string(),
            description: None,
            tags: Some(vec!["done".to_string()]),
            due_date: DueDate::Day(NaiveDate::from_ymd_opt(year, 10, 15).unwrap()),
            priority: 0,
            state: State::Done,
            line_number: 1,
            ..Default::default()
        };
        assert_eq!(res, expected);
    }

    #[test]
    fn test_parse_task_only_state() {
        let mut input = "- [ ]";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected = Task {
            subtasks: vec![],
            name: String::new(),
            description: None,
            tags: None,
            due_date: DueDate::NoDate,
            priority: 0,
            state: State::ToDo,
            line_number: 1,
            filename: String::new(),
            is_today: false,
        };
        assert_eq!(res, expected);
    }
    #[test]
    fn test_parse_task_with_due_date_words() {
        let mut input = "- [ ] today 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected_date = chrono::Local::now().date_naive();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = DueDate::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_with_weekday() {
        let mut input = "- [ ] monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();

        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = DueDate::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_with_weekday_this() {
        let mut input = "- [ ] this monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = DueDate::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_with_weekday_next() {
        let mut input = "- [ ] next monday 15:30 task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let now = chrono::Local::now();
        let expected_date = now
            .date_naive()
            .checked_add_days(Days::new(
                8 - u64::from(now.date_naive().weekday().number_from_monday()),
            ))
            .unwrap();
        let expected_time = NaiveTime::from_hms_opt(15, 30, 0).unwrap();
        let expected_due_date = DueDate::DayTime(NaiveDateTime::new(expected_date, expected_time));
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_without_due_date() {
        let mut input = "- [ ] task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        let expected_due_date = DueDate::NoDate;
        assert_eq!(res.due_date, expected_due_date);
    }

    #[test]
    fn test_parse_task_with_invalid_state() {
        let mut input = "- [invalid] task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_err());
    }

    #[test]
    fn test_parse_task_without_state() {
        let mut input = "task_name";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_err());
    }

    #[test]
    fn test_parse_task_with_invalid_priority() {
        let mut input = "- [ ] task_name p-9";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert_eq!(res.priority, 0);
    }

    #[test]
    fn test_parse_task_without_name() {
        let mut input = "- [ ]";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert_eq!(res.name, ""); // Default name is used when no name is provided
    }
    #[test]
    fn test_parse_task_with_today_flag() {
        let mut input = "- [ ] @t";
        let config = TasksConfig::default();
        let res = parse_task(&mut input, String::new(), &config);
        assert!(res.is_ok());
        let res = res.unwrap();
        assert!(res.is_today);
    }
}
