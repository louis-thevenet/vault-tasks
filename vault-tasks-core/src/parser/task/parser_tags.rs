use winnow::{combinator::preceded, token::take_while, PResult, Parser};

use super::token::Token;

/// Parses tags of the form "#tag".
pub fn parse_tag(input: &mut &str) -> PResult<Token> {
    let tag = preceded(
        '#',
        take_while(1.., ('_', '0'..='9', 'A'..='Z', 'a'..='z', '0'..='9')),
    )
    .parse_next(input)?;
    Ok(Token::Tag(tag.to_string()))
}

#[cfg(test)]
mod tests {
    use crate::parser::task::{parser_tags::parse_tag, token::Token};

    #[test]
    fn test_parse_tag_sucess() {
        let mut with_tag = "#test";
        assert_eq!(parse_tag(&mut with_tag), Ok(Token::Tag("test".to_string())));
    }
    #[test]
    fn test_parse_tag_symbols() {
        let mut with_tag = "#test_underscore123";
        assert_eq!(
            parse_tag(&mut with_tag),
            Ok(Token::Tag("test_underscore123".to_string()))
        );
    }
    #[test]
    fn test_parse_tag_fail() {
        let mut without_tag = "test";
        assert!(parse_tag(&mut without_tag).is_err());
    }
}
