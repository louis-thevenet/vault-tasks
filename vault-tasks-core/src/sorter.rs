use std::cmp::Ordering;

use chrono::NaiveTime;
use lexical_sort::lexical_cmp;
use strum::EnumIter;
use strum_macros::FromRepr;

use super::task::{DueDate, Task};

#[derive(Default, Clone, Copy, FromRepr, EnumIter, strum_macros::Display)]
pub enum SortingMode {
    #[default]
    #[strum(to_string = "Due Date")]
    ByDueDate,
    #[strum(to_string = "Title")]
    ByName,
}

impl SortingMode {
    pub fn next(self) -> Self {
        match self {
            SortingMode::ByDueDate => SortingMode::ByName,
            SortingMode::ByName => SortingMode::ByDueDate,
        }
    }
    pub fn sort(tasks: &mut [Task], sorter: SortingMode) {
        tasks.sort_by(|t1, t2| Self::cmp(t1, t2, sorter));
    }

    /// Compare two tasks by due date
    fn cmp_due_date(t1: &Task, t2: &Task) -> Ordering {
        match (&t1.due_date, &t2.due_date) {
            (DueDate::Day(d1), DueDate::Day(d2)) => d1.cmp(d2),
            (DueDate::DayTime(d1), DueDate::DayTime(d2)) => d1.cmp(d2),
            (DueDate::Day(d1), DueDate::DayTime(d2)) => d1.and_time(NaiveTime::default()).cmp(d2),
            (DueDate::DayTime(d1), DueDate::Day(d2)) => d1.cmp(&d2.and_time(NaiveTime::default())),
            (DueDate::NoDate, DueDate::Day(_) | DueDate::DayTime(_)) => Ordering::Greater,
            (DueDate::Day(_) | DueDate::DayTime(_), DueDate::NoDate) => Ordering::Less,
            _ => Ordering::Equal,
        }
    }
    /// Compares two tasks with the specified sorting mode
    /// Sorting mode is used first
    /// If equal, other attribues will be used:
    /// - State: `ToDo` < `Done` (in Ord impl of `State`)
    /// - The other sorting mode
    /// - Priority: usual number ordering
    /// - Tags: not used
    fn cmp(t1: &Task, t2: &Task, sorter: SortingMode) -> Ordering {
        let res_initial_sort = match sorter {
            SortingMode::ByDueDate => Self::cmp_due_date(t1, t2),
            SortingMode::ByName => lexical_cmp(&t1.name, &t2.name),
        };

        if !matches!(res_initial_sort, Ordering::Equal) {
            return res_initial_sort;
        }

        // Compare states
        let res = t1.state.cmp(&t2.state);
        if !matches!(res, Ordering::Equal) {
            return res;
        }

        // We do the other sorting methods
        let res = match sorter {
            SortingMode::ByDueDate => lexical_cmp(&t1.name, &t2.name),
            SortingMode::ByName => Self::cmp_due_date(t1, t2),
        };
        if !matches!(res, Ordering::Equal) {
            return res;
        }

        t1.priority.cmp(&t2.priority)
    }
}
#[cfg(test)]
mod tests {

    use insta::{assert_debug_snapshot, with_settings};

    use super::SortingMode;
    use crate::{parser::task::parse_task, task::Task, TasksConfig};
    #[test]
    fn task_sort_by_name() {
        let mut source = [
            "- [ ] test 10/11",
            "- [ ] test 10/9",
            "- [ ] test 10/10 p5",
            "- [ ] test 10/10 10:00",
            "- [x] zèbre",
            "- [x] zzz",
            "- [ ] zzz",
            "- [ ] test 10/10 p2",
            "- [x] test",
            "- [ ] test2",
            "- [ ] test 10/10 5:00",
            "- [ ] abc",
        ];
        let config = TasksConfig {
            use_american_format: true,
            task_state_marker: 'x',
            ..Default::default()
        };
        let mut tasks: Vec<Task> = source
            .iter_mut()
            .map(|input| parse_task(input, String::new(), &config).unwrap())
            .collect();

        let sorting_mode = SortingMode::ByName;
        SortingMode::sort(&mut tasks, sorting_mode);

        let tasks = tasks
            .iter()
            .map(|task| task.get_fixed_attributes(&config, 2))
            .collect::<Vec<String>>();

        with_settings!({
            info=>&source,
            description => "", // the template source code
        }, {
                assert_debug_snapshot!(tasks);
        });
    }
    #[test]
    fn task_sort_by_due_date() {
        let mut source = [
            "- [ ] test 10/11",
            "- [ ] test 10/9",
            "- [ ] test 10/10 p5",
            "- [ ] test 10/10 10:00",
            "- [x] zèbre",
            "- [x] zzz",
            "- [ ] zzz",
            "- [ ] test 10/10 p2",
            "- [x] test",
            "- [ ] test2",
            "- [ ] test 10/10 5:00",
            "- [ ] abc",
        ];
        let config = TasksConfig {
            use_american_format: true,
            task_state_marker: 'x',
            ..Default::default()
        };
        let mut tasks: Vec<Task> = source
            .iter_mut()
            .map(|input| parse_task(input, String::new(), &config).unwrap())
            .collect();

        let sorting_mode = SortingMode::ByDueDate;
        SortingMode::sort(&mut tasks, sorting_mode);

        let tasks = tasks
            .iter()
            .map(|task| task.get_fixed_attributes(&config, 2))
            .collect::<Vec<String>>();

        with_settings!({
            info=>&source,
            description => "", // the template source code
        }, {
                assert_debug_snapshot!(tasks);
        });
    }
}
