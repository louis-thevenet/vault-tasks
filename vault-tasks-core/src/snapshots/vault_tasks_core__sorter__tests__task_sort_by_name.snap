---
source: vault-tasks-core/src/sorter.rs
description: ""
expression: tasks
info:
  - " test 10/11"
  - " test 10/9"
  - " test 10/10 p5"
  - " test 10/10 10:00"
  - " zèbre"
  - " zzz"
  - " zzz"
  - " test 10/10 p2"
  - " test"
  - " test2"
  - " test 10/10 5:00"
  - " abc"
---
[
    "  - [ ] abc",
    "  - [ ] test 2024/10/09",
    "  - [ ] test 2024/10/10 p2",
    "  - [ ] test 2024/10/10 p5",
    "  - [ ] test 2024/10/10 05:00:00",
    "  - [ ] test 2024/10/10 10:00:00",
    "  - [ ] test 2024/10/11",
    "  - [x] test",
    "  - [ ] test2",
    "  - [x] zèbre",
    "  - [ ] zzz",
    "  - [x] zzz",
]
